src/uct/ib/rc/accel/rc_mlx5.inl

```diff
@@ -1820,3 +1847,196 @@ uct_rc_mlx5_iface_common_atomic_data(unsigned opcode, unsigned size, uint64_t va
+{
+    uct_ib_iface_recv_desc_t *desc;
+    void *hdr;
+
+    UCT_TL_IFACE_GET_RX_DESC(&iface->super.super.super, &iface->super.rx.mp,
+                             desc, return UCS_ERR_NO_MEMORY);
+    hdr           = uct_ib_iface_recv_desc_hdr(&iface->super.super, desc);
+    seg->srq.desc = desc;
+
+    seg->srq.ptr_mask |= UCS_BIT(UCT_IB_RX_SG_TL_HEADER_IDX);
+    seg->dptr[UCT_IB_RX_SG_TL_HEADER_IDX].lkey = htonl(desc->header_lkey);
+    seg->dptr[UCT_IB_RX_SG_TL_HEADER_IDX].addr = htobe64((uintptr_t)hdr);
+
+    VALGRIND_MAKE_MEM_NOACCESS(hdr,
+                               uct_ib_iface_tl_hdr_length(&iface->super.super));
+    return UCS_OK;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+uct_rc_mlx5_iface_common_seg_set_sge_payload_entry(
+        uct_rc_mlx5_iface_common_t *iface, uct_ib_mlx5_srq_seg_t *seg)
+{
+    uct_base_iface_t *base_iface   = &iface->super.super.super;
+    uct_ib_iface_recv_desc_t *desc = seg->srq.desc;
+
+    desc->payload = uct_rc_mlx5_rx_allocator_get_buffer(base_iface);
+    if (ucs_unlikely(desc->payload == NULL)) {
+        return UCS_ERR_NO_MEMORY;
+    }
+    desc->payload_lkey = uct_ib_memh_get_lkey(
+            base_iface->rx_allocator.cache.memh);
+
+    seg->srq.ptr_mask |= UCS_BIT(UCT_IB_RX_SG_PAYLOAD_IDX);
+    seg->dptr[UCT_IB_RX_SG_PAYLOAD_IDX].lkey = htonl(desc->payload_lkey);
+    seg->dptr[UCT_IB_RX_SG_PAYLOAD_IDX].addr = htobe64(
+            (uintptr_t)desc->payload);
+
+    VALGRIND_MAKE_MEM_NOACCESS(desc->payload,
+                               base_iface->rx_allocator.payload_length);
+    return UCS_OK;
+}
+
+static UCS_F_ALWAYS_INLINE ucs_status_t
+uct_rc_mlx5_iface_common_srq_set_seg_sge(uct_rc_mlx5_iface_common_t *iface,
+                                         uct_ib_mlx5_srq_seg_t *seg)
+{
+    uct_base_iface_t *base_iface = &iface->super.super.super;
+    uint64_t desc_map;
+    ucs_status_t status;
+
+    desc_map = ~seg->srq.ptr_mask & UCS_MASK(UCT_IB_RECV_SG_LIST_LEN);
+    if (!desc_map) {
+        return UCS_OK;
+    }
+
+    if (uct_rc_mlx5_rx_allocator_is_empty(base_iface)) {
+        status = uct_rc_mlx5_rx_allocator_get_buffers(base_iface);
+        if (ucs_unlikely(status != UCS_OK)) {
+            return status;
+        }
+    }
+
+    if (desc_map & UCS_BIT(UCT_IB_RX_SG_TL_HEADER_IDX)) {
+        status = uct_rc_mlx5_iface_common_seg_set_sge_header_entry(iface, seg);
+        if (ucs_unlikely(status != UCS_OK)) {
+            return status;
+        }
+    }
+
+    if (desc_map & UCS_BIT(UCT_IB_RX_SG_PAYLOAD_IDX)) {
+        /*
+         * No need to check for the return status here.
+         * RX Allocator cache contains at least one free buffer.
+         * We know that because before entering this branch we filled
+         * the cache with new buffers and checked the return status.
+         */
+        ucs_assertv(!uct_rc_mlx5_rx_allocator_is_empty(base_iface),
+                    "RX Allocator cache is empty");
```

Comment:
why need assert if uct_rc_mlx5_iface_common_seg_set_sge_payload_entry already returns UCS_ERR_NO_MEMORY in this case